symengine/llvm_double.h

```diff
@@ -0,0 +1,432 @@
+        fpm->add(llvm::createInstructionCombiningPass());
+        // Reassociate expressions.
+        fpm->add(llvm::createReassociatePass());
+        // Eliminate Common SubExpressions.
+        fpm->add(llvm::createGVNPass());
+        // Simplify the control flow graph (deleting unreachable blocks, etc).
+        fpm->add(llvm::createCFGSimplificationPass());
+        fpm->add(llvm::createPartiallyInlineLibCallsPass());
+        fpm->add(llvm::createLoadCombinePass());
+        fpm->add(llvm::createInstructionSimplifierPass());
+        fpm->add(llvm::createMemCpyOptPass());
+        fpm->add(llvm::createMergedLoadStoreMotionPass());
+        fpm->add(llvm::createBitTrackingDCEPass());
+        fpm->add(llvm::createAggressiveDCEPass());
+
+        fpm->doInitialization();
+
+        std::vector<llvm::Type *> inp;
+        for (int i = 0; i < 2; i++) {
+            inp.push_back(
+                llvm::PointerType::get(llvm::Type::getDoubleTy(*context), 0));
+        }
+        llvm::FunctionType *function_type = llvm::FunctionType::get(
+            llvm::Type::getVoidTy(*context), inp, false);
+        auto F = llvm::Function::Create(
+            function_type, llvm::Function::InternalLinkage, "", module.get());
+        F->setCallingConv(llvm::CallingConv::C);
+        {
+            llvm::SmallVector<llvm::AttributeSet, 4> attrs;
+            llvm::AttributeSet PAS;
+            {
+                llvm::AttrBuilder B;
+                B.addAttribute(llvm::Attribute::ReadOnly);
+                B.addAttribute(llvm::Attribute::NoCapture);
+                PAS = llvm::AttributeSet::get(module->getContext(), 1U, B);
+            }
+
+            attrs.push_back(PAS);
+            {
+                llvm::AttrBuilder B;
+                B.addAttribute(llvm::Attribute::NoCapture);
+                PAS = llvm::AttributeSet::get(module->getContext(), 2U, B);
+            }
+
+            attrs.push_back(PAS);
+            {
+                llvm::AttrBuilder B;
+                B.addAttribute(llvm::Attribute::NoUnwind);
+                B.addAttribute(llvm::Attribute::UWTable);
+                PAS = llvm::AttributeSet::get(module->getContext(), ~0U, B);
+            }
+
+            attrs.push_back(PAS);
+
+            F->setAttributes(
+                llvm::AttributeSet::get(module->getContext(), attrs));
+        }
+
+        // Add a basic block to the function. As before, it automatically
+        // inserts
+        // because of the last argument.
+        llvm::BasicBlock *BB
+            = llvm::BasicBlock::Create(*context, "EntryBlock", F);
+
+        // Create a basic block builder with default parameters.  The builder
+        // will
+        // automatically append instructions to the basic block `BB'.
+        builder = llvm::make_unique<llvm::IRBuilder<>>(BB);
+        builder->SetInsertPoint(BB);
+        auto fmf = llvm::FastMathFlags();
+        // fmf.setUnsafeAlgebra();
```

Comment:
I haven't added the fast-math flags here. Un-commenting this line would add it. I'll check to see whether this affects performance.